Update the documentation around the gocounter examples (#275)
redhat-et/bpfd: docs/developer/example-bpf.md

[examples/go-tc-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter).
These examples and the associated documentation is intended to provide the basics on how to deploy
and manage a BPF program using bpfd.
There are two ways to deploy these example applications, simply run locally on one machine, or deploy
to multiple nodes in a Kubernetes cluster.

The `go-xdp-counter` and  `go-tc-counter` examples each contain a BPF Program written in C
([xdp_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter/bpf/xdp_counter.c) and
[tc_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter/bpf/tc_counter.c))
that is compiled into BPF bytecode.
The BPF program counts packets that are received on an interface and stores the packet and byte
counts in a map that is accessible by the userspace portion.
The `go-xdp-counter` and  `go-tc-counter` examples each also have a userspace portion written in GO.
When run locally, the userspace program makes gRPC calls to `bpfd` requesting `bpfd` to load the BPF program
at the requested hook Point (XDP hook point or the TC hook point).

> When run in a Kubernetes deployment, the `bpfd-agen`t makes gRPC calls to `bpfd` requesting `bpfd` to load

Reviewer: Super Nit
```suggestion
When run in a Kubernetes deployment, the `bpfd-agent` makes gRPC calls to `bpfd` requesting `bpfd` to load
```

# Example BPF Programs

> Example applications that uses the `bpfd-go` bindings can be found in

Reviewer: ```suggestion
Example applications that use the `bpfd-go` bindings can be found in
```

> Example applications that uses the `bpfd-go` bindings can be found in

Author: Done

[examples/go-xdp-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter) and
[examples/go-tc-counter/](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter).

> These examples and the associated documentation is intended to provide the basics on deploying a BPF

Reviewer: ```suggestion
These examples and the associated documentation is intended to provide the basics on how to deploy and manage a BPF program using bpfd.
```

> These examples and the associated documentation is intended to provide the basics on deploying a BPF

Author: Done

program using bpfd with a go gRPC client.

> There are two ways to deploy these example applications, simply run locally, or run in a Kubernetes

Author: Done

cluster.

The `go-xdp-counter` and  `go-tc-counter` examples each contain a BPF Program written in C
([xdp_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-xdp-counter/bpf/xdp_counter.c) and
[tc_counter.c](https://github.com/redhat-et/bpfd/tree/main/examples/go-tc-counter/bpf/tc_counter.c))
that is compiled into BPF bytecode.
The BPF program counts packets that are received on an interface and stores the packet and byte
counts in a map that is accessible by the userspace portion.

> The `go-xdp-counter` and  `go-tc-counter` examples each also have a userspace portion written in GO that is

Author: Rewritten

responsible for loading the BPF program at the XDP hook point or the TC hook point, via gRPC calls to `bpfd`, and
then every 3 seconds polls the BPF map and logs the current counts.
The userspace code is leveraging the [cilium/ebpf library](https://github.com/cilium/ebpf)
to manage the maps shared with the BPF program.

> The programs are very similar, just varies where in the Linux networking stack the programs are inserted.

Author: Done

#### Prerequisites

**Assuming bpfd is already installed and running on the system**

1. All [requirements defined by the `cilium/ebpf` package](https://github.com/cilium/ebpf#requirements)
2. libbpf development package to get the required bpf c headers

    Fedora:

    `sudo dnf install libbpf-devel`

    Ubuntu:

    `sudo apt-get install libbpf-dev`

3. Cilium's `bpf2go` binary

    `go install github.com/cilium/ebpf/cmd/bpf2go@master`

#### Building Locally

To build the C based BPF counter bytecode, run:

```console
    cd src/bpfd/examples/go-xdp-counter/
    go generate
```

To build the Userspace GO Client run:

```console
    cd src/bpfd/examples/go-xdp-counter/
    go build
```

Repeat for TC if desired:
```console
    cd src/bpfd/examples/go-tc-counter/
    go generate
    go build
```

### Running On Host

The most basic way to deploy this example is running directly on a host system.
First, start or ensure `bpfd` is up and running.
[tutorial.md](../admin/tutorial.md) will guide you through deploying `bpfd`.
In all the examples of running `go-xdp-counter` or `go-tc-counter` on a host system,

> a bpfd-client certificate is used that is generated by bpfd.

Author: Done

The diagram below shows `go-xdp-counter` example, but the `go-tc-counter` example
operates exactly the same way.

![go-xdp-counter On Host](../img/gocounter-on-host.png)

Following the diagram (Purple numbers):
1. When `go-xdp-counter` userspace is started, it will send a gRPC request

> over mTLS to `bpfd` requesting `bpfd` to load the `go-xdp-counter` BPF bytecode located at

Author: Done

> `src/bpfd/src/examples/go-xdp-counter/bpf_bpfel.o` at a priority of 50 and on interface `ens3`.

Reviewer: ```suggestion
`bpfd/examples/go-xdp-counter/bpf_bpfel.o` at a priority of 50 and on interface `ens3`.
```

Author: Good catch. Done

These values are configurable as we will see later, but for now we will use the defaults
(except interface, which is required to be entered).

> 2. `bpfd` will load it's `dispatcher` BPF program, which links to the `go-xdp-counter` BPF program

Reviewer: ```suggestion
2. `bpfd` will load it's `dispatcher` BPF program, which links to the `go-xdp-counter` BPF program and return a UUID describing the running program.
```

Author: Done

    2022/12/02 15:59:41 580 bytes received

    :
```

Use `bpfctl` to show the `go-xdp-counter` BPF bytecode was loaded.

```console
    bpfctl list
    UUID                                  Type  Name   Path                                                      Metadata
    b6b2107c-f1a3-48ac-a145-1073c0979ba4  xdp   stats  /home/$USER/src/bpfd/examples/go-xdp-counter/bpf_bpfel.o  { "priority": 50, "iface": "ens3", "position": 0, "proceed_on": ["pass", "dispatcher_return"] }
```

Finally, press `<CTRL>+c` when finished with `go-xdp-counter`.

```console
    :

    2022/12/02 16:00:56 64 packets received
    2022/12/02 16:00:56 9280 bytes received

    2022/12/02 16:00:59 64 packets received
    2022/12/02 16:00:59 9280 bytes received

    ^C2022/12/02 16:01:00 Exiting...
    2022/12/02 16:01:00 Unloading Program: b6b2107c-f1a3-48ac-a145-1073c0979ba4
```

#### Running Unprivileged

To run the `go-xdp-counter` and `go-tc-counter` examples unprivileged (without `sudo`), the
following three steps must be performed.

##### Step 1: Create `bpfd` User Group

The [tutorial.md](../admin/tutorial.md) guide describes the different modes `bpfd` can be run in.

> Specifically, the [Systemd Service](../admin/tutorial.md#systemd-service) section describe how to

Reviewer: ```suggestion
Specifically, the [Systemd Service](../admin/tutorial.md#systemd-service) section explains how to
```

Author: Done

> start `bpfd` the `bpfd` Users and `bpfd` User Group.

Author: Done

> `bpfd` must be started as a Systemd Service and `go-xdp-counter` must be run from a User that is a

Reviewer: All this about the shared Socket is stale right? We got rid of passing the map FD over a socket

Author: I didn't realize it was gone. Removed documentation on it and also the const in both main.c files

    d541af30-69be-44cf-9397-407ee512547a
```

Then run the `go-xdp-counter` userspace program, passing in the UUID:

```console
    ./go-xdp-counter -iface ens3 -uuid d541af30-69be-44cf-9397-407ee512547a
    2022/12/02 17:01:38 Using Input: Interface=ens3 Source=d541af30-69be-44cf-9397-407ee512547a
    2022/12/02 17:01:41 180 packets received
    2022/12/02 17:01:41 26100 bytes received

    2022/12/02 17:01:44 184 packets received
    2022/12/02 17:01:44 26680 bytes received

    ^C2022/12/02 17:01:46 Exiting...
```

Then use `bpfctl` to unload the BPF bytecode:

```console
    bpfctl unload d541af30-69be-44cf-9397-407ee512547a
```

## Deploying On Kubernetes

This section will describe loading bytecode on a Kubernetes cluster and launching the userspace
program.
The approach is slightly different when running on a Kubernetes cluster.
The BPF bytecode should be loaded by an administrator, not the userspace program itself.

![gocounter On Kubernetes](../img/gocounter-on-k8s.png)

### Loading BPF Bytecode On Kubernetes

This assumes there is already a Kubernetes cluster running and `bpfd` is running in the cluster

> (see [How to Manually Deploy bpfd on Kubernetes](../admin/k8s-deployment.md)).

Reviewer: This like points to a super stake doc currently.... (I'm assuming what @anfredette Is doing will refactor it?)

Author: Yes, that's my assumption.

Instead of using the userspace program or `bpfctl` to load the BPF bytecode as done above, the bytecode

> will be loaded by creating a `BpfProgramConfig` CRD. Edit the

Author: Done

```

Then push images to a remote repository:

```console
    docker login quay.io
    docker push quay.io/$USER/go-xdp-counter:latest
    docker push quay.io/$USER/go-tc-counter:latest
```

#### Loading A Userspace Container Image

The userspace program in a Kubernetes Deployment no longer interacts directly with `bpfd`.
Instead, the userspace program running on each node reads the `BpdProgram` to determine the
map location.
For example, the output above shows the maps as:

```console
    kubectl get bpfprograms go-xdp-counter-example-bpfd-deployment-worker -o yaml
    :
    spec:
      programs:
        30f8264a-ba07-4adc-9824-85a2c755e85e:
    :
          maps:
            xdp_stats_map: /run/bpfd/fs/maps/30f8264a-ba07-4adc-9824-85a2c755e85e/xdp_stats_map
    :
```

To interact with the KubeApiServer, RBAC must be setup properly to access the `BpdProgram`
object.
The `bpfd-operator` defined the yaml for several ClusterRoles that can be used to access the
different `bpfd` related CRD objects with different access rights.
The example userspace containers will use the `bpfprogram-viewer-role`, which allows Read-Only
access to the `BpfProgram` object.

> This ClusterRole needs to be created.

Author: Thanks. Removed commands to create and delete.

     namespace "go-xdp-counter" deleted
     serviceaccount "bpfd-app-go-xdp-counter" deleted
     clusterrolebinding.rbac.authorization.k8s.io "bpfd-app-rolebinding-go-xdp-counter" deleted
     daemonset.apps "go-xdp-counter-ds" deleted

    kubectl delete -f examples/go-xdp-counter/kubernetes-deployment/go-xdp-counter-bytecode.yaml
     bpfprogramconfig.bpfd.io "go-xdp-counter-example" deleted

    kubectl delete -f examples/go-tc-counter/kubernetes-deployment/go-tc-counter.yaml
     namespace "go-tc-counter" deleted
     serviceaccount "bpfd-app-go-tc-counter" deleted
     clusterrolebinding.rbac.authorization.k8s.io "bpfd-app-rolebinding-go-tc-counter" deleted
     daemonset.apps "go-tc-counter-ds" deleted

    kubectl delete -f examples/go-tc-counter/kubernetes-deployment/go-tc-counter-bytecode.yaml
     bpfprogramconfig.bpfd.io "go-tc-counter-example" deleted

    kubectl delete -f bpfd-operator/config/rbac/ebpfprogram_viewer_role.yaml
     clusterrole.rbac.authorization.k8s.io "bpfprogram-viewer-role" deleted
```

## Notes

Notes regarding this document:
* Source of images used in this document can be found in

>   [bpfd Upstream Images](https://docs.google.com/presentation/d/1wU4xu6xeyk9cB3G-Nn-dzkf90j1-EI4PB167G7v-Xl4/edit?usp=sharing).

Author: No, worked for me. It's on my Red Hat account, so need to be logged in as Red Hat.
